execution/executor-types/src/lib.rs

```diff
@@ -443,12 +451,25 @@ impl StateComputeResult {
                     _ => None,
                 }
             })
-            .chain((!self.has_reconfiguration()).then_some(Transaction::StateCheckpoint(block_id)))
+            .chain(
+                (!self.has_reconfiguration()).then_some(self.block_end_info.clone().map_or(
+                    Transaction::StateCheckpoint(block_id),
+                    |block_end_info| {
+                        Transaction::BlockEpilogue(BlockEpiloguePayload::WithBlockEndInfo {
+                            block_id,
+                            block_end_info,
+                        })
+                    },
+                )),
+            )
             .collect::<Vec<_>>();
 
         assert!(
             self.has_reconfiguration()
-                || matches!(output.last(), Some(Transaction::StateCheckpoint(_))),
+                || matches!(
+                    output.last(),
+                    Some(Transaction::StateCheckpoint(_) | Transaction::BlockEpilogue(_))
```

Comment:
nit: I tend to suggest adding a method for this simple logic, maybe `.is_non_reconfig_block_ending()`